app/_src/gateway/reference/wasm.md

```diff
@@ -0,0 +1,141 @@
+---
+title: WebAssembly
+content_type: reference
+badge: free
+---
+
+# WebAssembly in Kong
+
+In addition to Lua plugins, {{site.base_gateway}} supports WebAssembly as a
+means of extending proxy functionality and performing custom business logic.
+This is made possible by Kong's [open source implementation](https://github.com/Kong/ngx_wasm_module) of
+the [Proxy-Wasm specification](https://github.com/proxy-wasm/spec).
+
+## Concepts
+
+### filter
+
+In the context of Proxy-Wasm, a "filter" refers to a specific component or
+module that is executed at some stage of request/response handling. Filters can
+perform actions such as:
+
+* read or write request or response headers
+* read or write the request or response body
+* terminate the request early and send a self-generated response
+
+If you are familiar with Kong's Lua plugin support, you can think of filters
+as semi-analogous to plugins.
+
+Filters differ from Lua plugins in their file format at runtime. A Lua plugin
+typically spans several different Lua source code files, all packaged and
+deployed in text format. Meanwhile, filters must be compiled from source code
+into a single binary file with the `.wasm` extension.
+
+In order to be usable in {{site.base_gateway}}, all filter files must be present
+on disk in a single directory, denoted by the `wasm_filters_path` configuration
+directive.
+
+### filter chain
+
+A filter chain is the database entity representing one or more filters executed
+for each request to a particular service or route, each one with its
+configuration.
+
+Example:
+
+```yaml
+---
+filter-chains:
+  name: my-filter-chain
+  # filter chains can be toggled on/off for testing
+  enabled: true
+  filters:
+    # this maps to $wasm_filters_path/my_filter.wasm
+    - name: my_filter
+      # at runtime, configuration is passed to the filter as a byte array
+      #
+      # filter configuration is inherently untyped--schema and
+      # encoding/serialization are left up to the filter code itself
+      config: foo=bar baz=bat
```

Comment:
I wasn't able to get Kong to accept this type of config. I am inclined to change this to the format below. What do you think @flrgh @hishamhm?

Reply:
(When I look at a line like this the first thing in my mind is to wonder if this needs quotes for it to work — my distrust of yaml runs deep!)

This format failed in your tests because the config format parser is implemented by the filter itself, and all filters we published in repos parse JSON only. But we have some filters in the test files for ngx_wasm_module which use this more "raw string" format, and even i've seen ones in the proxy-wasm-go-sdk testsuite where the config itself was just a number like `1` — proxy-wasm doesn't specify a format.

I didn't even know the `>-` magic syntax below, but if that produces a nested string that doesn't need quoting, that's better than all the `config: "{\"foo\": \"bar\"}"` stuff we've been generally doing for JSON-based configs...

I guess the point of this example was to show that raw configs like this _are_ possible (and some third-party filters out there probably do it). I'd just quote like this to make it look less magical, I guess:

```suggestion
      config: "foo=bar baz=bat"
```

Reply:
What error are you getting?

Reply:
@hishamhm you are right, I am still not used to filters validating the config :)

@flrgh when initializing the filter, this is the type of error we get: `[warn] 2309#0: [proxy-wasm]["proxy_wasm_rust_my_filter" #0] on_configure: failed parsing configuration: my_greeting="hello": Invalid type`